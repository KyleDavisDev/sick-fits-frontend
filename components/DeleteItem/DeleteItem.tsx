import * as React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import ErrorMessage from "../ErrorMessage/ErrorMessage";
import { ALL_ITEMS_QUERY } from "../Items/Items";

export const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

export interface IDeleteItemProps {
  id: string;
  skip: number;
  first: number;
}

export interface IDeleteItemState {}

export default class DeleteItem extends React.Component<
  IDeleteItemProps,
  IDeleteItemState
> {
  constructor(props: IDeleteItemProps) {
    super(props);

    this.state = {};
  }

  public render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {deleteItem => {
          return (
            <button
              onClick={() => {
                if (confirm("Are you sure you want to delete this item?")) {
                  deleteItem().catch(err => {
                    alert(err.message);
                  });
                }
              }}
            >
              Delete Item
            </button>
          );
        }}
      </Mutation>
    );
  }

  public update = (cache, payload) => {
    if (cache.data.data.ROOT_QUERY) {
      // manually update the cache on the client, so it matches the server
      // 1. Read the cache for the items we want
      const data = cache.readQuery({
        query: ALL_ITEMS_QUERY,
        variables: { skip: this.props.skip, first: this.props.first }
      });

      // 2. Filter the deleted itemout of the page
      data.items = data.items.filter(
        item => item.id !== payload.data.deleteItem.id
      );

      // 3. Put the items back!
      cache.writeQuery({
        query: ALL_ITEMS_QUERY,
        data,
        variables: { skip: this.props.skip, first: this.props.first }
      });
    }
  };
}
