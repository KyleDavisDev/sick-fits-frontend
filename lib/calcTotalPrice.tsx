export default function(items) {
  return items.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
