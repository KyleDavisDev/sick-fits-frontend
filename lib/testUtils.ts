import casual from "casual";
import { ReactWrapper } from "enzyme";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  created: casual.unix_time,
  id: casual.uuid,
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = (overrides?: any) => ({
  __typename: "CartItem",
  id: casual.uuid,
  created: casual.unix_time,
  quantity: casual.integer(0, 15),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

const fakeCart = (overrides?: any) => ({
  __typename: "Cart",
  id: casual.uuid,
  items: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
  user: fakeUser(),
  created: casual.unix_time,
  updated: casual.unix_time,
  isActive: true,
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  store: any;
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeCart,
  fakeOrder,
  fakeOrderItem,
  simulateInputChange
};
