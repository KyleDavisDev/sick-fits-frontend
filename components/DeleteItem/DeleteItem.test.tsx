import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { ApolloConsumer } from "react-apollo";
import DeleteItem, { DELETE_ITEM_MUTATION } from "./DeleteItem";
import { fakeUser, simulateInputChange } from "../../lib/testUtils";

const fakeId = "string";
const fakeSkip = 3;
const fakeFirst = 3;

describe("<DeleteItem />", () => {
  it("renders and matches snapshot", () => {
    const wrapper = mount(
      <MockedProvider>
        <DeleteItem id={fakeId} first={fakeFirst} skip={fakeSkip} />
      </MockedProvider>
    );

    // check render
    expect(wrapper).toBeTruthy();

    // check snapshot
    expect(wrapper.find("button")).toMatchSnapshot();
  });
});
