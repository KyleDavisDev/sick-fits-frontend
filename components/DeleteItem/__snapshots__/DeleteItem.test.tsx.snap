// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteItem /> renders and matches snapshot 1`] = `
<button
  onClick={[Function]}
>
  Delete Item
</button>
`;
